fix(replset): fix set protocolVersion=1 for replicaset configuration in MongoDB 3.2+

addisonElliott opened the pull request:
Replicaset configuration has two protocol versions. *protocolVersion* 0 is the original. *protocolVersion* 1 was added in MongoDB 3.2 and is **required** in MongoDB 4.0+ [source](https://docs.mongodb.com/manual/reference/replica-set-protocol-versions/).

*protocolVersion* field is missing in this library. This PR adds the *protocolVersion* and allows the user to manually specify it. The default is to set the protocolVersion to 1 if using MongoDB 3.2+, otherwise set to 0.

This is the error message that appears if you try to run this library with MongoDB 4.0+:
```
MongoError: Support for replication protocol version 0 was removed in MongoDB 4.0. Downgrade to MongoDB version 3.6 and upgrade your protocol version to 1 before upgrading your MongoDB version
```

**Note:** When running the tests, I received some error if the protocolVersion was not set to 1 for MongoDB 3.2+. I don't remember exactly but it had to do with the Shard configuration servers.

mbroadst commented:
Hm, I'm not against adding in a concept of specifying the protocol version (and in fact, should just default to 1 if above 2.6), but I don't believe that this doesn't work for MongoDB 4.x+. I use this library every day to run tests on my local machine against v4.1.5+, and we have a TravisCI job for the node driver that tests against 4.0.x. Do you have an example of a build failure?

addisonElliott commented:
Yeah, here's it occurring on my Windows machine (just ran it now). Here's the version of MongoDB I'm using.

Seems to only occur at "add new member to set" and not the replicaset tests that come beforehand.

![Screenshot 1](https://i.imgur.com/spUc6zq.png)
![Screenshot 2](https://i.imgur.com/IUW2h6I.png)

addisonElliott commented:
Yeah, the failure of the one test causes the remaining ones to fail. I was thinking the same thing after seeing your comment.

Did not realize you had a work environment with potentially other MongoDB instances running. I just had in mind TravisCI and local machines. Will make the changes tonight.

Should I change the logic of this PR to set protocolVersion to 1 if running MongoDB 2.6+?

mbroadst commented:
@addisonElliott yeah sorry that was my mistake, was about to update here. It's safe to assume `protocolVersion: 1` starting with 3.2. It's worth adding a way to set this explicitly from the user's perspective, but unlikely that people will actually be using PV0.

addisonElliott commented:
@mbroadst There is a way for the developer to specify the `protocolVersion` via the options. I also added JSDoc string for the `ReplSet` constructor while I was adding options.

I've updated the title of this PR to follow the angular JS convention as you said. One thing that I was unsure about from reading the guide, should the commits within the PR be squashed & follow the same convention or is that something the maintainer will handle when merging?

mbroadst commented on lib/replset.js:
```diff
@@ -60,8 +80,12 @@ class ReplSet extends EventEmitter {
     this.replSet = options.replSet;
 
+    // Remove the values from the options
+    delete this.options['protocolVersion'];
```
We don't usually mutate objects like this because of how V8 works, also I don't see a reason it would hurt to leave it in the options.

addisonElliott replied:
@mbroadst The answer to this comment and the 4th comment is that I was just trying to stay consistent. In the constructor, values are taken out from `options` and saved locally. The `options` are then passed as `clientOptions` to each `Server` instance ([here](https://github.com/mongodb-js/mongodb-topology-manager/blob/master/lib/replset.js#L80)). I did a search in the class and there are **no** cases of using values from options besides the constructor.

Here's three spots in the constructor where the values are extracted from options (and in 2/3 of the cases, they are deleted from the options):
* https://github.com/mongodb-js/mongodb-topology-manager/blob/master/lib/replset.js#L28-L29
* https://github.com/mongodb-js/mongodb-topology-manager/blob/master/lib/replset.js#L47-L56
* https://github.com/mongodb-js/mongodb-topology-manager/blob/master/lib/replset.js#L63

My original guess was because the `options` are passed to `Server` and the person wanted to prevent accidental repeat options, but if that doesn't matter to you, then we don't need to do that. Maybe it was to allow the user to easily access these options from the class (i.e. `ReplSet.configSettings` vs `ReplSet.options.configSettings`)

Would you like me to change these occurrences as well (`replSet`, `electionCycleWaitMS`, `configSettings`, & `retryWaitMS`) to stay consistent? (Change them to read directly from options and not store locally)

Also, can you give a brief explanation of what you mean by **because of how V8 works** (interested in learning why)

mbroadst replied:
I think the motto for working on these legacy projects is "Don't touch it unless you're already working on the piece of code, or writing new code." So in this case, I would say let's not pull it out and follow old convention, but also let's not fix the old instances unless we're specifically working on those features. 

Regarding "how v8 works", I think this [blog](https://v8.dev/blog/fast-properties) might have the relevant information. Generally, you want to research v8's "hidden classes"

mbroadst commented on lib/replset.js:
```diff
@@ -158,6 +182,15 @@ class ReplSet extends EventEmitter {
+        // Set protocol version for replicaset to be given value in options or
+        // default to version 1 if using MongoDB 3.2 or above
+        // Protocol version 1 is supported since 3.2+ and required for 4.0+
+        var protocolVersion = self.protocolVersion;
```
while we're at it: generally prefer `let` and `const` (so called lexically scoped variables) to `var`. Its not something that we go out of our way to change, but instead change when working in a piece of code or writing new code.

addisonElliott replied:
👍

mbroadst commented on lib/replset.js:
```diff
@@ -158,6 +182,15 @@ class ReplSet extends EventEmitter {
+        var protocolVersion = self.protocolVersion;
+        if (typeof protocolVersion !== 'number') {
+          protocolVersion =
+            result.version[0] >= 4 || (result.version[0] === 3 && result.version[1] >= 2) ? 1 : 0;
```
Would it make sense to introduce a dependency on `semver` so that we can make server version checks more readable? We already depend on it in e.g. `mongodb-version-manager` so this wouldn't really affect the whole dependency graph that much

mbroadst replied:
:+1:

mbroadst commented on lib/replset.js:
```diff
@@ -167,6 +200,7 @@ class ReplSet extends EventEmitter {
         var config = generateConfiguration(
           self.replSet,
           self.version,
+          protocolVersion,
```
This is fine, but I wonder if it's actually useful to save off `protocolVersion` anymore, and why we aren't just passing the `options` right to `generateConfiguration`?

addisonElliott replied:
I misunderstood this comment the first time. I just did it for 'cleaner code' in my opinion. I can give that a whirl and see how it looks though.

addisonElliott commented:
@mbroadst Found a bug in eslint/prettier. Have a double ternary operator causes issue with the spacing. In Windows it wants the second ternary operator to be indented, while Linux wants the opposite for both ternary operators to be inline. Going to look into it...

mbroadst commented:
Ah, looks like another case where we might want to introduce that dependency on `semver` for readability. But also I think nested ternarys lead to less readable code, could we just unroll it?
